apps/ingest-from-bucket/app/provider.py

```diff
@@ -0,0 +1,163 @@
+        try:
+            import json
+            from google.oauth2 import service_account
+
+            account_json = json.loads(self.credentials)
+            creds = service_account.Credentials.from_service_account_info(account_json)
+            self.client = self.storage.Client( credentials=creds)
+        except Exception as e:
+            logger.error(f"Failed to create client for gs: {e}")
+            return False
+        try:
+            r=self.client.list_buckets(max_results=10)
+        except Exception as e:
+            logger.error(f"Failed to read buckets for gs: {e}")
+            return False
+
+        try:
+            r=self.client.lookup_bucket(self.bucket)
+        except Exception as e:
+            logger.error(f"Could not get bucket info ({self.bucket}) for gs: {e}")
+            return False
+        return True
+
+    def scan(self,filter_func):
+        load_count = 0
+        paths = []
+
+        blobs = self.client.list_blobs(self.bucket) 
+        for blob in blobs:
+            if filter_func(blob.name):
+                load_count = load_count + 1
+                paths.append(blob.name)
+
+        return paths
+
+class AWSProvider(Provider):
+    def __init__(self,bucket,access_key,secret_key):
+        super(AWSProvider,self).__init__(bucket)
+        import boto3
+        self.client = \
+        boto3.client('s3',aws_access_key_id=access_key,aws_secret_access_key=secret_key)
+        # store this here because some things might be picky about region.
+        self.region_name = self.client.meta.region_name
+    def url_column_name(self):
+        return "s3_url"
+    def url_scheme_name(self):
+        return "s3:"
+    def verify(self):
+
+        # Verify Account
+        try:
+            self.client.list_buckets(MaxBuckets=10)
+        except Exception as e:
+            #InvalidAccessKey - bad AK (short/wrong)
+            #SignatureDoesNotMatch - bad SK
+            logger.error(f"Failed to list buckets for s3: {e}")
+            return False
+
+        # Verify Bucket
+
+        try:
+            res = self.client.head_bucket(Bucket=self.bucket)
+        except Exception as e:
+            # 404 not found if bad bucket name
+            logger.error(f"Could not get bucket info ({self.bucket}) for s3: {e}")
+            return False
+        return True
+
+    def scan(self,filter_func):
+
+        load_count = 0
+        paths = []
+
+        object_paginator = self.client.get_paginator('list_objects_v2')
+        bucket_iter = object_paginator.paginate(Bucket=self.bucket)
+        for page in bucket_iter:
+            for obj in page['Contents']:
+                name = obj['Key']
+                if filter_func(name):
+                    load_count = load_count + 1
+                    paths.append(name)
+
+        return paths
+
+
+
+# potential errors to pay attention to:
+# bad access key
+# bad secret key
+# expired key
+# bad bucket name
+# no access to bucket
```

Comment:
I once ran into a situation, where I was not able to access contents of a public bucket using the regular gs client.
The bucket I tried it was ad-demos-datasets, in GCP. Would this Workflow work with such public buckets?